File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/reactiveobject/UaReadVariableValueTransaction.java
package org.opcfoundation.webserver.service.transactions.reactiveobject;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.DataValue;
import org.eclipse.milo.opcua.stack.core.types.builtin.DateTime;
import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.enumerated.TimestampsToReturn;
import org.opcfoundation.webserver.addressspace.reactiveobject.UaReactiveObjectType;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManagerReactiveObject;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.service.transactions.base.UaReadTransaction;
import org.opcfoundation.webserver.types.common.UaObjectId;
import org.opcfoundation.webserver.types.common.UaObjectIdentifier;
import org.opcfoundation.webserver.types.common.UaChildId;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadVariableValueRequest;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadVariableValueResponse;
import org.opcfoundation.webapi.service.types.ReadContext;

import java.util.*;
import java.util.concurrent.CompletableFuture;

public class UaReadVariableValueTransaction extends UaReadTransaction {
    private final UaObjectIdentifier objectId;
    private final List<UaChildId> childIds;
    private final NodeManagerReactiveObject nodeManager;

    public UaReadVariableValueTransaction(
            ReadContext context,
            UaObjectIdentifier objectId,
            Map<Integer, UaChildId> handleIdsAndChildIds,
            NodeManagerReactiveObject nodeManager)
    {
        super(context,new ArrayList<>(handleIdsAndChildIds.keySet()));
        this.objectId = objectId;
        this.childIds = new ArrayList<>(handleIdsAndChildIds.values());
        this.nodeManager = nodeManager;
    }

    public CompletableFuture<Void> execute()
    {
        // System.out.println("Read variable of object " + objectId);

        try {
            UaReactiveObjectType objectType = nodeManager.findObjectType(objectId);
            if (null == objectType) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

            Set<UaChildId> variableIds = new HashSet<>(childIds);
            UaObject instanceDeclaration = nodeManager.findInstanceDeclaration(objectId);

            ReadVariableValueRequest request = new ReadVariableValueRequest(
                    new UaObjectId(
                            objectId.getId(),
                            instanceDeclaration),
                    variableIds);

            return objectType.onReadVariablesValue(request).
                        thenAccept(this::setResults).
                        exceptionally(ex -> buildErrorResponse(ex.getCause()));
        } catch (Exception e) {
            buildErrorResponse(e);
        }

        return CompletableFuture.completedFuture(null);
    }

    private void setResults(ReadVariableValueResponse response)
    {
        DateTime sourceTimestamp = DateTime.NULL_VALUE;
        DateTime serverTimestamp = DateTime.NULL_VALUE;

        if (TimestampsToReturn.Source == timestampsToReturn)
        {
            sourceTimestamp = DateTime.now();
        } else if (TimestampsToReturn.Server == timestampsToReturn) {
            serverTimestamp =  DateTime.now();
        } else if (TimestampsToReturn.Both == timestampsToReturn) {
            sourceTimestamp = DateTime.now();
            serverTimestamp = sourceTimestamp;
        }

        for (UaChildId item: childIds)
        {
            DataValue result = response.getResults().get(item);

            if (null != result)
            {
                results.add(new DataValue(
                        result.value(),
                        result.statusCode(),
                        (result.statusCode().isBad()) ? DateTime.NULL_VALUE : sourceTimestamp,
                        (result.statusCode().isBad()) ? DateTime.NULL_VALUE : serverTimestamp));
            } else {
                results.add(new DataValue(Variant.NULL_VALUE, StatusCode.of(StatusCodes.Bad_NotReadable), null));
            }
        }
    }
}

File: opcua-webservice-node/src/server/service/transactions/reactiveobject/UaReadVariableValueTransaction.ts
import { StatusCodes, TimestampsToReturn } from "opcua-webapi";
import { makeUaStatusCode, UaDataValue, UaError, UaVariant } from "opcua-webapi-ts";
import { ReadContext } from "../../..";
import { NodeManagerReactiveObject } from "../../../addressspace/nodemanager/NodeManagerReactiveObject";
import { UaChildId, UaObjectId, UaObjectIdentifier } from "../../../types";
import { ReadVariableValueRequest, ReadVariableValueResponse } from "../../message";
import { UaReadTransaction } from "../base/UaReadTransaction";

export class UaReadVariableValueTransaction extends UaReadTransaction {
    private readonly objectId: UaObjectIdentifier;
    private readonly childIds: Array<UaChildId>;
    private readonly nodeManager: NodeManagerReactiveObject;

    constructor(
        context: ReadContext,
        objectId: UaObjectIdentifier,
        handleIdsAndChildIds: Map<number, UaChildId>,
        nodeManager: NodeManagerReactiveObject,
    ) {
        super(context, [...handleIdsAndChildIds.keys()]);
        this.objectId = objectId;
        this.childIds = [...handleIdsAndChildIds.values()];
        this.nodeManager = nodeManager;
    }

    override async execute(): Promise<void> {
        try {
            const objectType = this.nodeManager.findObjectType(this.objectId);
            if (objectType === null) {
                throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
            }

            const variableIds = new Set(this.childIds.map((childId) => childId.toString()));
            const instanceDeclaration = this.nodeManager.findInstanceDeclaration(this.objectId);
            const request = new ReadVariableValueRequest(
                new UaObjectId(this.objectId.id, instanceDeclaration),
                variableIds,
            );

            const response = await objectType.onReadVariablesValue(request);
            this.setResults(response);
        } catch (error) {
            this.buildErrorResults(error);
        }
    }

    private setResults(response: ReadVariableValueResponse): void {
        let sourceTimestamp: Date | undefined;
        let serverTimestamp: Date | undefined;
        const now = new Date();

        if (this.timestampsToReturn === TimestampsToReturn.Source) {
            sourceTimestamp = now;
        } else if (this.timestampsToReturn === TimestampsToReturn.Server) {
            serverTimestamp = now;
        } else if (this.timestampsToReturn === TimestampsToReturn.Both) {
            sourceTimestamp = now;
            serverTimestamp = now;
        }

        this._results = [];

        for (const item of this.childIds) {
            const result = response.results.get(item.toString());

            if (result !== undefined) {
                const hasBadStatus = result.statusCode.isBad();
                this._results.push(new UaDataValue(
                    result.value,
                    result.statusCode,
                    hasBadStatus ? undefined : sourceTimestamp,
                    hasBadStatus ? undefined : serverTimestamp,
                ));
            } else {
                this._results.push(
                    new UaDataValue(
                        UaVariant.null(),
                        makeUaStatusCode(StatusCodes.BadNotReadable),
                    ),
                );
            }
        }
    }
}